File: frontend/src/lib/types/retrieval.ts
/** Matches backend ``RetrievalMethod``. */
export type RetrievalMethod = "vector" | "bm25" | "hybrid_fusion" | "multi_query";

export interface RetrievedChunkResult {
  chunk_id: string;
  text: string;
  document_id: string;
  filename: string;
  file_type: string;
  source: string;
  page_number: number;
  section: string | null;
  chunk_index: number;
  chunking_strategy: string;
  score: number;
  retrieval_method: RetrievalMethod;
}

export interface PipelineStage {
  id: string;
  label: string;
  enabled: boolean;
  executed: boolean;
  result_count: number | null;
  details: Record<string, unknown>;
}

export interface RetrievalConfiguration {
  query_transformation_enabled: boolean;
  multi_query_enabled: boolean;
  hybrid_search_enabled: boolean;
  context_optimization_enabled: boolean;
  reranking_enabled: boolean;
}

export interface RetrievalExploreRequest {
  query: string;
  top_k?: number;
  document_ids?: string[];
  filenames?: string[];
  file_types?: string[];
  sections?: string[];
}

export interface RetrievalExploreResponse {
  query: string;
  retrieval_query: string;
  generated_queries: string[] | null;
  configuration: RetrievalConfiguration;
  pipeline: PipelineStage[];
  vector_results: RetrievedChunkResult[];
  bm25_results: RetrievedChunkResult[];
  fused_results: RetrievedChunkResult[] | null;
  results: RetrievedChunkResult[];
  metadata: Record<string, unknown>;
}

export type RetrievalResultView = "final" | "vector" | "bm25" | "fused";

export type RetrievalSortField = "score" | "filename" | "page_number";
export type RetrievalSortDirection = "asc" | "desc";

export type RetrievalMethodFilter = "all" | RetrievalMethod;

File: app/schemas/retrieval.py
"""Retrieval explorer API schemas."""

from __future__ import annotations

from typing import Any, Literal

from pydantic import BaseModel, Field

from app.schemas.query import QueryRequest


class RetrievalExploreRequest(QueryRequest):
    """Request body for ``POST /retrieval/explore``."""


RetrievalMethod = Literal["vector", "bm25", "hybrid_fusion", "multi_query"]


class RetrievedChunkResponse(BaseModel):
    """A retrieved chunk exposed for retrieval exploration."""

    chunk_id: str
    text: str
    document_id: str
    filename: str
    file_type: str
    source: str
    page_number: int
    section: str | None = None
    chunk_index: int
    chunking_strategy: str
    score: float
    retrieval_method: RetrievalMethod


class PipelineStageResponse(BaseModel):
    """One stage in the retrieval pipeline."""

    id: str
    label: str
    enabled: bool
    executed: bool
    result_count: int | None = None
    details: dict[str, Any] = Field(default_factory=dict)


class RetrievalConfigurationResponse(BaseModel):
    """Active retrieval configuration flags."""

    query_transformation_enabled: bool
    multi_query_enabled: bool
    hybrid_search_enabled: bool
    context_optimization_enabled: bool
    reranking_enabled: bool = False


class RetrievalExploreResponse(BaseModel):
    """Raw retrieval pipeline output for the Retrieval Explorer."""

    query: str
    retrieval_query: str
    generated_queries: list[str] | None = None
    configuration: RetrievalConfigurationResponse
    pipeline: list[PipelineStageResponse]
    vector_results: list[RetrievedChunkResponse] = Field(default_factory=list)
    bm25_results: list[RetrievedChunkResponse] = Field(default_factory=list)
    fused_results: list[RetrievedChunkResponse] | None = None
    results: list[RetrievedChunkResponse] = Field(default_factory=list)
    metadata: dict[str, Any] = Field(default_factory=dict)
